docs/reference/configuration/configuration.md

```diff
@@ -1,48 +1,276 @@
+  - `value`: an optional field that defines the value of the parameter.
+  - `required`: an optional key that specifies whether the parameter must have a value before compiling the code (`false` by default). It's not possible to compile a source tree with one or more required parameters that don't have a value. Generally, setting `required` to true is only useful when `value` is not set.
+  - `macro_name`: an optional name for the macro defined at compile time for this configuration parameter. The configuration system automatically figures out the corresponding macro name for a configuration parameter, but the user can override this automatically computed name by specifying `macro_name`.
+
+You define a macro by value by using an integer or string instead of the description object, such as `param3` above. Defining a parameter by value is equivalent to defining it with a description object with the key `value` set to the value in place of the description object, the key `help` not set, the key `macro_name` not set and the key `required` set to `false`. Defining a parameter by value is equivalent to a configuration parameter defined with a description object with the key `value` set to the value in place of the description object, the key `help` unset, the key `macro_name` unset, and the key `required` set to `false`.
+
+<span class="notes">**Note:** The name of a parameter in `config` can't contain a dot (`.`) character.</span>
+
+The configuration system appends a prefix to the name of each parameter, so a parameter with the same name in a library does not conflict with parameters of the same name in targets or other libraries. The prefix is:
+
+| Location | Prefix |
+| -------- | ------ |
+| Target | `target.` |
+| Any library | The name of the library, as found in the `name` section of `mbed_lib.json`, followed by a dot (.) |
+| Application | `app.` |
 
-The default stack size of a user spawned thread is 4 kilobytes. You can configure the stack size of user spawned threads by using the [Mbed config system](/docs/development/tools/index.html#the-configuration-system) or passing in the [Thread constructor](/docs/development/reference/thread.html). You can reduce the default stack size for user spawned threads from 4K to 2K in the following two ways:
+### `mbed_lib.json` format specification
 
-1. Create an `mbed_app.json` file for your project with the following content:
+`mbed_lib.json` is a JSON formatted document that contains a root JSON Object. The keys within this object are sections. See the allowed sections and their meanings below:
 
-    ```JSON
-    {
-        "config": {
-            "thread-stack-size": {
-                "value": 2000
-            }
+| Section | Required | Meaning |
+| ------- | -------- | ------- |
+| `name`  | Yes      | Name of the library. Must be unique. May not be `app` or `target`.|
+| `macros` | No | List of macros to define on the command-line. |
```

Comment:
Are they actually set on the command-line and not `mbed_config.h`? If so, are they are any important considerations other than just to look for them in a different place when building? (if no, might be helpful to just mention that distinction so understand its just a simple relocation vs behavior change)

Reply:
If you decide not to keep the "section" idea, be sure to change it in this table column title, too.

Though its use here makes me question if I understood correctly above that a "section" necessarily refers to the topmost keys in the JSON file.

Reply:
Sorry I missed this. They're in `mbed_config.h`

Reply:
@bmcdonnell-ionx Section is used as the topmost keys in the JSON file here too. Is there something that made you question that notion?

Reply:
I was thinking topmost within the whole app hierarchy, i.e. within `mbed_app.json`. I guess you mean topmost within any JSON file, then?

Reply:
You're guess is correct. How could I make that more clear?

Reply:
I'm not sure that it needs clarification. I hadn't looked at it in context for a while.